src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs

```diff
@@ -128,27 +131,72 @@ fn has_significant_drop_in_scrutinee<'tcx, 'a>(cx: &'a LateContext<'tcx>, expr:
 
 struct SigDropHelper<'a, 'tcx> {
     cx: &'a LateContext<'tcx>,
-    returns: bool,
+    has_significant_temporary: bool,
     is_chain_end: bool,
+    seen_types: FxHashSet<Ty<'tcx>>,
 }
 
 impl<'a, 'tcx> SigDropHelper<'a, 'tcx> {
     fn new(cx: &'a LateContext<'tcx>) -> SigDropHelper<'a, 'tcx> {
         SigDropHelper {
             cx,
-            returns: false,
+            has_significant_temporary: false,
             is_chain_end: true,
+            seen_types: FxHashSet::default(),
         }
     }
+
     fn get_expr_ty(&self, ex: &'tcx Expr<'_>) -> Ty<'tcx> {
         self.cx.typeck_results().expr_ty(ex)
     }
+
+    fn has_seen_type(&mut self, ty: Ty<'tcx>) -> bool {
+        if self.seen_types.contains(&ty) {
+            true
+        } else {
+            self.seen_types.insert(ty);
+            false
+        }
```

Comment:
```suggestion
        !self.seen_types.insert(ty)
```

`insert` will return `false` if it already contains this item. So this function can be simplified.

Reply:
Much nicer! Thank you!